### ISSUE TITLE
Issue with C++ Turbo Module in React Native 0.76.5 (Works in iOS, Not in Android)

### ISSUE THREAD

**Issue by cloud2303** [facebook/react-native#48298]
### Description

I am using React Native 0.76.5, and I followed the steps to create a basic C++ Turbo Module demo. However, it does not work on Android, though it works fine on iOS. When I downgrade to React Native 0.76.1, the demo works on both iOS and Android.

Please advise on what might be causing this issue in 0.76.5.
I also came across [this pull request](https://github.com/facebook/react-native/pull/48156#issue-2723478104), but even after switching to the latest version of OnLoad.cpp, the demo still doesn’t work on 0.76.5. My example repository is using 0.76.1, and I can reproduce the issue by switching to 0.76.5.

### Steps to reproduce

1. npm install

### React Native Version

0.76.5

### Affected Platforms

Runtime - Android

### Areas

TurboModule - The New Native Module System

### Output of `npx react-native info`

```text
System:
  OS: Windows 11 10.0.22631
  CPU: (12) x64 12th Gen Intel(R) Core(TM) i5-12400
  Memory: 558.32 MB / 15.72 GB
Binaries:
  Node:
    version: 20.17.0
    path: C:\Program Files\nodejs\node.EXE
  Yarn: Not Found
  npm:
    version: 10.8.2
    path: C:\Program Files\nodejs\npm.CMD
  Watchman: Not Found
SDKs:
  Android SDK:
    API Levels:
      - "30"
      - "34"
      - "35"
    Build Tools:
      - 30.0.2
      - 34.0.0
      - 35.0.0
    System Images:
      - android-35 | Pre-Release 16 KB Page Size Google Play ARM Intel x86_64
        Atom
    Android NDK: Not Found
  Windows SDK:
    AllowDevelopmentWithoutDevLicense: Enabled
    AllowAllTrustedApps: Enabled
    Versions:
      - 10.0.19041.0
      - 10.0.20348.0
      - 10.0.22621.0
IDEs:
  Android Studio: Not Found
  Visual Studio:
    - 17.9.34701.34 (Visual Studio Professional 2022)
Languages:
  Java:
    version: 17.0.6
    path: /c/Users/fujiahui/.jdks/jbr-17.0.6/bin/javac
  Ruby: Not Found
npmPackages:
  "@react-native-community/cli":
    installed: 15.0.1
    wanted: 15.0.1
  react:
    installed: 18.3.1
    wanted: 18.3.1
  react-native:
    installed: 0.76.1
    wanted: 0.76.1
  react-native-windows: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: true
iOS:
  hermesEnabled: Not found
  newArchEnabled: Not found
```

### Stacktrace or Logs

```text
dont load the custom cpp module
```

### Reproducer

https://github.com/cloud2303/new-arch-cpp-module-test

**Comment by sarthak-d11** [comment]
Hey @cloud2303 
Would it be possible for you to share a reproducible example using React Native version 0.76.5? As the issue is occurring with this version, and having a sample would be very helpful for further investigation.

**Comment by cloud2303** [comment]
Hi  @sarthak-d11 
I have adjusted the React Native version to 0.76.5, and you can find the updated repository at https://github.com/cloud2303/new-arch-cpp-module-test. This should provide a reproducible example of the issue.

**Comment by jankosecki** [comment]
Getting the same issue - I've started by upgrading our app from 0.74 to 0.76.5 - once the app built but crashed due to:
```
Invariant Violation: TurboModuleRegistry.getEnforcing(...): 'NativeModule' could not be found.
```
I tried to simplify the app and I followed step-by-step [guide](https://reactnative.dev/docs/the-new-architecture/pure-cxx-modules) which resulted in the same error anyway.

After finding this thread, I've downgraded RN to 0.76.1 in the sample app and my main app and both work fine on 0.76.1.

Upgrading to 0.76.2 breaks the app again so whatever causes the issue, must have been introduced in 0.76.2 release.

**Comment by cipolleschi** [comment]
I tried to investigate this a bit, and it looks like that, after the commit shared by @jankosecki, the `OnLoad` provided by the app is completely ignored by the app. I think we need to customize the CMakeLists.txt provided in the gradle.build
```groovy
    externalNativeBuild {
        cmake {
            path "src/main/jni/CMakeLists.txt"
        }
    }
```

I tried something, like adding the `OnLoad.cpp` to the `CMakeLists.txt` file but we get duplicated symbols.

I don't have a lot more context here and I'm afraid we have to wait for Nicola to be back after PTO, he should be back next year.
